library/worldgen/material_rule/src/testmod/java/org/quiltmc/qsl/worldgen/material_rule/QuiltMaterialRuleTest.java

```diff
@@ -0,0 +1,129 @@
+		SurfaceRules.MaterialCondition blueNoise2 = SurfaceRules.noiseThreshold(NoiseParametersKeys.CALCITE, 0.25, 0.30);
+
+		SurfaceRules.MaterialRule LIGHT_BLUE_CONCRETE = SurfaceRules.block(Blocks.LIGHT_BLUE_CONCRETE.getDefaultState());
+		SurfaceRules.MaterialRule PINK_CONCRETE = SurfaceRules.block(Blocks.PINK_CONCRETE.getDefaultState());
+		SurfaceRules.MaterialRule WHITE_CONCRETE = SurfaceRules.block(Blocks.WHITE_CONCRETE.getDefaultState());
+
+		context.materialRules().add(0,
+				SurfaceRules.condition(
+						SurfaceRules.abovePreliminarySurface(),
+						SurfaceRules.condition(
+								SurfaceRules.ON_FLOOR,
+								SurfaceRules.sequence(
+										SurfaceRules.condition(blueNoise1, LIGHT_BLUE_CONCRETE),
+										SurfaceRules.condition(pinkNoise1, PINK_CONCRETE),
+										SurfaceRules.condition(whiteNoise, WHITE_CONCRETE),
+										SurfaceRules.condition(pinkNoise2, PINK_CONCRETE),
+										SurfaceRules.condition(blueNoise2, LIGHT_BLUE_CONCRETE)
+								)
+						)
+				)
+		);
+	}
+
+	@Override
+	public void modifyNetherRules(SurfaceMaterialRuleContext.Nether context) {
+		SurfaceRules.MaterialCondition redNoise = SurfaceRules.noiseThreshold(NoiseParametersKeys.NETHER_STATE_SELECTOR, -0.04, -0.08);
+		SurfaceRules.MaterialCondition orangeNoise = SurfaceRules.noiseThreshold(NoiseParametersKeys.NETHER_STATE_SELECTOR, -0.8, -0.12);
+		SurfaceRules.MaterialCondition yellowNoise = SurfaceRules.noiseThreshold(NoiseParametersKeys.NETHER_STATE_SELECTOR, -0.12, -0.16);
+		SurfaceRules.MaterialCondition greenNoise = SurfaceRules.noiseThreshold(NoiseParametersKeys.NETHER_STATE_SELECTOR, -0.16, -0.20);
+		SurfaceRules.MaterialCondition lightBlueNoise = SurfaceRules.noiseThreshold(NoiseParametersKeys.NETHER_STATE_SELECTOR, -0.20, -0.24);
+		SurfaceRules.MaterialCondition darkBlueNoise = SurfaceRules.noiseThreshold(NoiseParametersKeys.NETHER_STATE_SELECTOR, -0.24, -0.28);
+		SurfaceRules.MaterialCondition purpleNoise = SurfaceRules.noiseThreshold(NoiseParametersKeys.NETHER_STATE_SELECTOR, -0.28, -0.32);
+
+		SurfaceRules.MaterialRule RED_CONCRETE = SurfaceRules.block(Blocks.RED_CONCRETE.getDefaultState());
+		SurfaceRules.MaterialRule ORANGE_CONCRETE = SurfaceRules.block(Blocks.ORANGE_CONCRETE.getDefaultState());
+		SurfaceRules.MaterialRule YELLOW_CONCRETE = SurfaceRules.block(Blocks.YELLOW_CONCRETE.getDefaultState());
+		SurfaceRules.MaterialRule GREEN_CONCRETE = SurfaceRules.block(Blocks.GREEN_CONCRETE.getDefaultState());
+		SurfaceRules.MaterialRule LIGHT_BLUE_CONCRETE = SurfaceRules.block(Blocks.LIGHT_BLUE_CONCRETE.getDefaultState());
+		SurfaceRules.MaterialRule DARK_BLUE_CONCRETE = SurfaceRules.block(Blocks.BLUE_CONCRETE.getDefaultState());
+		SurfaceRules.MaterialRule PURPLE_CONCRETE = SurfaceRules.block(Blocks.PURPLE_CONCRETE.getDefaultState());
+
+		context.materialRules().add(0,
+				SurfaceRules.condition(
+						SurfaceRules.UNDER_CEILING,
+						SurfaceRules.sequence(
+								SurfaceRules.condition(redNoise, RED_CONCRETE),
```

Comment:
General complaint: double tabssssssssss; I hate 'em
I'd say "remove 'em", but i'm not so sure if this would fit the style rules